contracts/Orchestrator.sol

```diff
@@ -46,35 +48,107 @@ contract Orchestrator is Ownable {
+
+        // perform external call
+        // todo: @thegostep solc v0.4 does not return revert string on call
+        // todo: @thegostep solc v0.4 does not support specifying gaslimit on call
+        // decide if upgrade solc or implement in assembly
+        // https://solidity.readthedocs.io/en/v0.4.24/units-and-global-variables.html#address-related
+        (bool success, bytes memory res) = address(transaction.destination).call(transaction.data);
+
+        // Check if any of the atomic transactions failed, if not, decode return data
+        bytes[] memory returnData;
+        if (!success) {
+            // parse revert message
+            (string memory revertMessage, bytes32 revertCode) = _getRevertMsg(res);
+            // if approved revert, log it and continue
+            if (transaction.revertOK[revertCode]) {
+                emit TransactionFailed(transaction.destination, index, transaction.data, revertMessage);
+            } 
+            // else revert batch
+            else {
+                revert("Transaction Failed");
+            }
+        } else {
+            // decode and return call return values
+            returnData = abi.decode(res, (bytes[]));
+        }
+
+        // explicit return
+        return (success, returnData);
+	}
+
+    /**
+     * @notice Get the revert message and code from a call.
+     * @param res bytes Response of the call.
+     * @return revertMessage string Revert message.
+     * @return revertCode bytes32 Revert code.
+     */
+	function _getRevertMsg(bytes memory res) internal pure returns (string memory revertMessage, bytes32 revertCode) {
+        // If there is no prefix to the revert reason, we know it was an OOG error
+        if (res.length == 0) {
+            revertMessage = "Transaction out of gas";
+        }
+		// If the revert reason length is less than 68, then the transaction failed silently (without a revert message)
+		else if (res.length < 68) {
+            revertMessage = "Transaction reverted silently";
+        }
+        // Else extract revert message
+        else {
+	        bytes memory revertData = res.slice(4, res.length - 4); // Remove the selector which is the first 4 bytes
+		    revertMessage = abi.decode(revertData, (string)); // All that remains is the revert string
+        }
+        // obtain revert code
+        revertCode = keccak256(revertMessage);
+        // explicit return
+        return (revertMessage, revertCode);
+	}
+
     /**
      * @notice Adds a transaction that gets called for a downstream receiver of rebases
      * @param destination Address of contract destination
      * @param data Transaction data payload
+     * @param gasLimit Transaction gas limit
+     * @param revertOKs Transaction approved revert codes
      */
-    function addTransaction(address destination, bytes data)
+    function addTransaction(address destination, bytes data, uint256 gasLimit, bytes32[] revertOKs)
```

Comment:
define policy of how to audit peripheral
- no upgrade
- no gas revert spoofing

Reply:
- a revert with no message would trigger the out of gas error check

Reply:
- avoid external call into solc <0.4